src/impute.jl

```diff
@@ -119,13 +126,28 @@ julia> [k => parent(parent(v)) for (k, v) in Impute.filter(ds; dims=:loc).data]
 """
 Impute.apply(ds::KeyedDataset, f::Filter; dims) = Impute.apply!(deepcopy(ds), f; dims=dims)
 
+_pattern(dims::Pattern) = dims
+_pattern(dims::Tuple) = Pattern(dims)
+_pattern(dims) = Pattern(:__, dims)
+
 function Impute.apply!(ds::KeyedDataset, f::Filter; dims)
+    pattern = _pattern(dims)
+    dim = pattern.segments[end]
+
+    dim in (:_, :__) && throw(ArgumentError(
+        "$pattern points to an ambiguous dimension in the dataset. " *
+        "Kwarg `dims` must end with a dimname."
+    ))
+    checkpaths = dimpaths(ds, pattern)
+
     # Limit our constraint map to paths containing the supplied dim
     cmap = filter(constraintmap(ds)) do (constraint, paths)
```

Comment:
Is `filter` meant to not filter anything that isn't attached to a constraint?

Reply:
Hmm, that's a good point. I don't see why we couldn't just run a second loop after over all `dimpaths` not covered by any existing constraints?

Reply:
Fixed!